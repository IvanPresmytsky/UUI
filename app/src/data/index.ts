import { svc } from '../services';

export * from './apiDefinition';
export type TAppContext = Awaited<ReturnType<typeof loadAppContext>>;
export async function loadAppContext() {
    if (!svc.api) {
        throw new Error('svc.api not available');
    }
    const [
        { content: navigation },
        { content: summaries },
    ] = await Promise.all([
        svc.api.getDocsGenExports(),
        svc.api.getDocsGenSummaries(),
    ]);
    return {
        docsGen: {
            navigation,
            summaries,
        },
    };
}
