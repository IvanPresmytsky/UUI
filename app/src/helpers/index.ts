export * from './copyTextToClipboard';
export * from './getQuery';
export * from './getCoreProps';
export * from './loremIpsum';
export * from './getCurrentTheme';
export * from './dayJsHelper';
