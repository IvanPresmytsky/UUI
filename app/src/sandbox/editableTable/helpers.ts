import { ItemsMap } from '@epam/uui-core';
import { Task } from './types';

const findAllChildren = (tasks: ItemsMap<number, Task>, parentTask: Task) => {
    const children: Task[] = [];
    tasks.forEach((task) => {
        if (task.parentId === parentTask.id) {
            children.push(task);
        }
    });
    let ids: number[] = [];
    children.forEach((task) => {
        ids.push(task.id);
        const innerChildren = findAllChildren(tasks, task);
        ids = ids.concat(innerChildren);
    });
    return ids;
};

export const deleteTaskWithChildren = (tasks: ItemsMap<number, Task>, taskToDelete: Task | null): ItemsMap<number, Task> => {
    let taskToBeDeleted = taskToDelete;
    if (taskToBeDeleted === undefined) {
        tasks.forEach((task) => {
            if (task.parentId === undefined) {
                taskToBeDeleted = task;
            }
        });
    }

    if (!taskToBeDeleted) {
        return tasks;
    }

    const childrenIds = findAllChildren(tasks, taskToBeDeleted);
    let newItemsMap = tasks;

    [taskToBeDeleted.id, ...childrenIds].forEach((id) => {
        newItemsMap = newItemsMap.set(id);
    });

    return newItemsMap;
};
