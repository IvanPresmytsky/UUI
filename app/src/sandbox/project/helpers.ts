import { getOrderBetween, maxOrderStr, minOrderStr } from '@epam/uui-core';
import { Task } from './types';

/**
 * Finds a string, which is placed alphabetically in desired position of the list.
 * If relativeTo is not specified, it is assumed that you need 'before' the first or 'after' the last item
 * If relativeTo is specified, location is before/after the relativeTo string.
 *  @see getOrderString function for details about order strings.
 */
export function getInsertionOrder(existingOrders: string[], position: 'before' | 'after', relativeTo?: string) {
    if (!relativeTo) {
        if (position === 'before') {
            // inserting at the top of the list, is the same as inserting after the minOrder
            relativeTo = minOrderStr;
            position = 'after';
        } else if (position === 'after') {
            // inserting at the bottom of the list, is the same as inserting before the maxOrder
            relativeTo = maxOrderStr;
            position = 'before';
        }
    }

    if (position === 'before') {
        let maxOrder = minOrderStr;
        existingOrders.forEach((order) => {
            if (order < relativeTo && order > maxOrder) {
                maxOrder = order;
            }
        });
        return getOrderBetween(maxOrder, relativeTo);
    } else {
        let minOrder = maxOrderStr;
        existingOrders.forEach((order) => {
            if (order > relativeTo && order < minOrder) {
                minOrder = order;
            }
        });
        return getOrderBetween(relativeTo, minOrder);
    }
}

const findAllChildren = (tasks: Task[], parentTask: Task) => {
    const children = tasks.filter((task) => task.parentId === parentTask.id);
    let ids: number[] = [];
    children.forEach((task) => {
        ids.push(task.id);
        const innerChildren = findAllChildren(tasks, task);
        ids = ids.concat(innerChildren);
    });
    return ids;
};

export const deleteTaskWithChildren = (tasks: Record<number, Task>, taskToDelete: Task | null): Record<number, Task> => {
    let taskToBeDeleted = taskToDelete;
    if (taskToBeDeleted === null) {
        const rootItems = Object.values(tasks).filter((task) => task.parentId === undefined);
        taskToBeDeleted = rootItems[rootItems.length - 1];
    }

    const childrenIds = findAllChildren(Object.values(tasks), taskToBeDeleted);
    [taskToBeDeleted.id, ...childrenIds].forEach((id) => {
        delete tasks[id];
    });

    return tasks;
};
