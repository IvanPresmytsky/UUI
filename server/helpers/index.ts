export * from './getFilterPredicate';
export * from './getOrderComparer';
export * from './indexToOrder';
export * from './generateTestData';
export * from './buildSearchTree';
export * from './getData';
