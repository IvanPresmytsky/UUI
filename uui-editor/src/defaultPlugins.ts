import { createJuicePlugin } from '@udecode/plate-juice';
import { createDeserializeDocxPlugin } from '@udecode/plate-serializer-docx';
import { isBlockAboveEmpty, isSelectionAtBlockStart } from '@udecode/plate-common';
import { ResetNodePlugin, createResetNodePlugin } from '@udecode/plate-reset-node';
import { createSoftBreakPlugin } from '@udecode/plate-break';
import {
    paragraphPlugin,
    PARAGRAPH_TYPE,
    QUOTE_PLUGIN_KEY,
    TODO_ELEMENT_KEY,
    noteTypes,
} from './plugins';

const resetBlockTypesCommonRule = {
    types: [
        QUOTE_PLUGIN_KEY,
        TODO_ELEMENT_KEY,
        ...noteTypes,
    ],
    defaultType: PARAGRAPH_TYPE,
};

const resetBlockTypePlugin: { options: Partial<ResetNodePlugin> } = {
    options: {
        rules: [
            {
                ...resetBlockTypesCommonRule,
                hotkey: 'Enter',
                predicate: isBlockAboveEmpty,
            },
            {
                ...resetBlockTypesCommonRule,
                hotkey: 'Backspace',
                predicate: isSelectionAtBlockStart,
            },
        ],
    },
};

/**
 * Please make sure defaultPlugins and all your plugins are not interfere
 * with the following list when disableCorePlugins prop hasn't been set
 * https://github.com/udecode/plate/blob/main/docs/BREAKING_CHANGES.md#general
 */
export const defaultPlugins: any = [
    createDeserializeDocxPlugin(), // depends on juice plugin
    createJuicePlugin(),
    paragraphPlugin(),
    createResetNodePlugin(resetBlockTypePlugin),
    createSoftBreakPlugin(),
];
