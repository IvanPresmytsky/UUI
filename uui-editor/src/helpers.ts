import {
    Range, Editor,
} from 'slate';
import {
    PlatePlugin, Value, createPlateEditor, getPlugins, useEditorState, createNode, PlateEditor, createPlugins,
} from '@udecode/plate-common';
import { createPlateUI } from './components';
import { PARAGRAPH_TYPE } from './plugins/paragraphPlugin';
import { migrateSlateSchema, SlateSchema } from './migrations/slate_migrations';
import { EditorValue } from './types';

export function getBlockDesirialiser(blockTags: Record<string, string>) {
    return (el: any, next: any) => {
        const block = blockTags[el.tagName.toLowerCase()];

        if (block) {
            return {
                object: 'block',
                type: block,
                nodes: next(el.childNodes),
            };
        }
    };
}

export function getMarkDeserializer(marks: Record<string, string>) {
    return (el: any, next: any) => {
        const mark = marks[el.tagName.toLowerCase()];

        if (mark) {
            return {
                object: 'mark',
                type: mark,
                nodes: next(el.childNodes),
            };
        }
    };
}

export function isTextSelected(editor: any, inFocus: boolean) {
    const { selection } = editor;
    return (selection && inFocus && !Range.isCollapsed(selection) && Editor.string(editor, selection) !== '');
}

export function isImageSelected(editor: any) {
    const { selection, getFragment } = editor;
    const node = getFragment()[0]?.type;
    return selection && node === 'image';
}

// TODO: get rid of that
export function useIsPluginActive(key: string): boolean {
    const editor = useEditorState();
    const plugins = getPlugins(editor);
    return plugins.some((plugin) => plugin.key === key);
}

export const isEditorValueEmpty = (value: Value) => {
    return (
        !value
        || (value.length === 0
        || (value.length === 1 && value[0].type === PARAGRAPH_TYPE && value[0].children[0].text === ''))
    );
};

export class SelectionUtils {
    static getSelection(params: { shadowRoot: ShadowRoot | undefined }) {
        const { shadowRoot } = params;
        if (shadowRoot) {
            if ((shadowRoot as any).getSelection) {
                // Chrome/Edge. See details here: https://caniuse.com/mdn-api_shadowroot_getselection
                return (shadowRoot as any).getSelection();
            }
        }
        // Works fine in other cases
        return window.getSelection();
    }

    static getSelectionRange0(params: { selection: Selection; shadowRoot: ShadowRoot | undefined }) {
        const { selection, shadowRoot } = params;
        if (shadowRoot) {
            if ((selection as any).getComposedRanges) {
                // Webkit. https://w3c.github.io/selection-api/#dom-selection-getcomposedranges
                return (selection as any).getComposedRanges(shadowRoot)[0];
            }
        }
        return selection.getRangeAt(0);
    }
}

export const createTempEditor = (plugins: PlatePlugin[]): PlateEditor => {
    return createPlateEditor({
        plugins: createPlugins((plugins).flat(), {
            components: createPlateUI(),
        }),
    });
};

/** Consider slate and plate migarions */
export const initializeEditor = (editor: PlateEditor<Value>, v: EditorValue): Value => {
    let value: Value;
    if (!v) {
        value = [createNode(PARAGRAPH_TYPE)];
    } else {
        if (!Array.isArray(v)) {
            value = migrateSlateSchema(v); // slate migraitons
        } else {
            value = v;
        }
    }

    editor.children = value;
    editor.normalize({ force: true }); // plate migratoins

    return editor.children;
};

/** type guard to support two modes: with strictNullCheck and without, try toggle editor tsconfig setting */
export const isSlateSchema = (value: EditorValue): value is SlateSchema => {
    return !!value && !Array.isArray(value);
};

export const isPlateValue = (value: EditorValue): value is Value => {
    return Array.isArray(value);
};
