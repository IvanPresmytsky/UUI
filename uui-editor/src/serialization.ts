import {
    PlatePlugin,
    createPlateEditor,
    createPlugins,
    deserializeHtml,
    parseHtmlDocument,
} from '@udecode/plate-common';
import { serializeHtml } from '@udecode/plate-serializer-html';
import { createPlateUI } from './components';
import { EditorValue } from './types';
import { migrateSchema } from './migration';
import {
    baseMarksPlugin,
    headerPlugin,
    superscriptPlugin,
    listPlugin,
    quotePlugin,
    linkPlugin,
    imagePlugin,
    videoPlugin,
    iframePlugin,
    codeBlockPlugin,
    paragraphPlugin,
} from './plugins';

const workingPlugins = [
    paragraphPlugin(),
    baseMarksPlugin(),
    headerPlugin(),
    superscriptPlugin(),
    listPlugin(),
    quotePlugin(),
    linkPlugin(),
    imagePlugin(),
    videoPlugin(),
    iframePlugin(),
    codeBlockPlugin(),
];

export const createTempEditor = (plugins?: PlatePlugin[]) => {
    return createPlateEditor({
        plugins: createPlugins((plugins || workingPlugins).flat(), {
            components: createPlateUI(),
        }),
    });
};

export const createDeserializer = () => {
    const editor = createTempEditor();

    return (data: string) => {
        const document = parseHtmlDocument(data);
        return deserializeHtml(editor, { element: document.body }) as EditorValue;
    };
};

export const createSerializer = () => {
    const editor = createTempEditor();

    return (value: EditorValue) => {
        return serializeHtml(editor, {
            nodes: migrateSchema(value),
        });
    };
};
