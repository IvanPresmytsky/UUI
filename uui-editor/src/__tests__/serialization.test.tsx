import fs from 'fs';
import path from 'path';
import { createSerializer } from '../serialization';
import {
    baseMarksPlugin,
    headerPlugin,
    colorPlugin,
    superscriptPlugin,
    listPlugin,
    toDoListPlugin,
    quotePlugin,
    linkPlugin,
    notePlugin,
    uploadFilePlugin,
    attachmentPlugin,
    imagePlugin,
    videoPlugin,
    iframePlugin,
    separatorPlugin,
    tablePlugin,
    codeBlockPlugin,
} from '../plugins';
import { defaultPlugins } from '../defaultPlugins';
import { createPlateEditor } from '@udecode/plate-core';
import { editorValueMock } from './data/md-serialization-content';
import { createTempEditor } from '../helpers';

export const readTestFile = (filepath: string): string => {
    const absoluteFilepath = path.resolve(__dirname, filepath);
    return fs.readFileSync(absoluteFilepath, 'utf8');
};

export const createClipboardData = (html: string, rtf?: string): DataTransfer =>
    ({
        getData: (format: string) => (format === 'text/html' ? html : rtf),
    } as any);

const plugins = [
    ...defaultPlugins,
    baseMarksPlugin(),
    headerPlugin(),
    colorPlugin(),
    superscriptPlugin(),
    listPlugin(),
    toDoListPlugin(),
    quotePlugin(),
    linkPlugin(),
    notePlugin(),
    uploadFilePlugin({}),
    attachmentPlugin(),
    imagePlugin(),
    videoPlugin(),
    iframePlugin(),
    separatorPlugin(),
    tablePlugin(),
    codeBlockPlugin(),
];

const initEditor = () => {
    const editor = createTempEditor(plugins);
    editor.children = [
        {
            data: {},
            type: 'paragraph',
            children: [
                {
                    text: '',
                },
            ],
        },
    ];
    return createPlateEditor({
        editor,
        plugins,
    });
};

describe('serialization', () => {
    it('should paste data with tables', async () => {
        const editor = initEditor();

        const data = createClipboardData(readTestFile('./data/html-tables-serialization.html'));
        editor.insertData(data);

        expect(editor.children).toMatchSnapshot();
    });

    it('should serialize markdown', async () => {
        const serializeMd = createSerializer('md');
        const serializedMd = serializeMd(editorValueMock);
        expect(serializedMd).toMatchSnapshot();
    });
});
