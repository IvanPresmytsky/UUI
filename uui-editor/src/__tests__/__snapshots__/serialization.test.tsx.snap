// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serialization should paste data with tables 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "text": "",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "text": "",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "background": "rgb(255, 255, 0)",
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "Sinple ",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
          Object {
            "background": "rgb(255, 255, 0)",
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "Table ",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
        ],
        "type": "table_row",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "With yellow ",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "back ",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
        ],
        "type": "table_row",
      },
    ],
    "type": "table",
  },
  Object {
    "children": Array [
      Object {
        "text": "",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "Lorem ipsum ",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "Lorem ipsum ",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "Lorem ipsum ",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
        ],
        "type": "table_row",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "1 ",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "In eleifend velit vitae libero sollicitudin euismod. ",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "Lorem ",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
        ],
        "type": "table_row",
      },
      Object {
        "children": Array [
          Object {
            "attributes": Object {
              "colspan": "4",
            },
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "2 ",
                  },
                ],
                "type": "paragraph",
              },
              Object {
                "children": Array [
                  Object {
                    "text": "Cras fringilla ipsum magna, in fringilla dui commodo a. ",
                  },
                ],
                "type": "paragraph",
              },
              Object {
                "children": Array [
                  Object {
                    "text": "Ipsum ",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
        ],
        "type": "table_row",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "3 ",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
          Object {
            "attributes": Object {
              "colspan": "2",
            },
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "Aliquam erat volutpat. ",
                  },
                ],
                "type": "paragraph",
              },
              Object {
                "children": Array [
                  Object {
                    "text": "Lorem ",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
        ],
        "type": "table_row",
      },
      Object {
        "children": Array [
          Object {
            "attributes": Object {
              "colspan": "4",
            },
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "4 ",
                  },
                ],
                "type": "paragraph",
              },
              Object {
                "children": Array [
                  Object {
                    "text": "Fusce vitae vestibulum velit. ",
                  },
                ],
                "type": "paragraph",
              },
              Object {
                "children": Array [
                  Object {
                    "text": "Lorem ",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
        ],
        "type": "table_row",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "5 ",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "Etiam vehicula luctus fermentum. ",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "Ipsum ",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "table_cell",
          },
        ],
        "type": "table_row",
      },
    ],
    "type": "table",
  },
]
`;
